debezium-embedded/src/main/java/io/debezium/embedded/AsyncEmbeddedEngine.java

```diff
@@ -0,0 +1,1266 @@
+        LOGGER.debug("Waiting max. for {} ms for individual source tasks to start.", taskStartupTimeout);
+        final int nTasks = tasks.size();
+        Exception error = null;
+        // To avoid leaked resources, we have to ensure that all tasks that were scheduled to start are really started before we continue with the execution in
+        // the main (engine) thread and change engine state. If any of the scheduled tasks has failed, catch the exception, wait for other tasks to start and then
+        // re-throw the exception and let engine stop already running tasks gracefully during the engine shutdown.
+        for (int i = 0; i < nTasks; i++) {
+            try {
+                final Future<Void> taskFuture = taskCompletionService.poll(taskStartupTimeout, TimeUnit.MILLISECONDS);
+                if (taskFuture != null) {
+                    taskFuture.get(); // we need to retrieve the results to propagate eventual exceptions
+                }
+                else {
+                    throw new InterruptedException("Time out while waiting for source task to start.");
+                }
+
+                LOGGER.debug("Started task #{} out of {} tasks.", i + 1, nTasks);
+            }
+            catch (Exception e) {
+                LOGGER.debug("Task #{} (out of {} tasks) failed to start. Failed with {}", i + 1, nTasks, e.getMessage());
+                if (LOGGER.isDebugEnabled()) {
+                    e.printStackTrace();
+                }
+
+                // Store only the first error.
+                if (error == null) {
+                    error = e;
+                }
+                continue;
+            }
+            LOGGER.debug("Calling connector callback after task is started.");
+            connectorCallback.ifPresent(DebeziumEngine.ConnectorCallback::taskStarted);
+        }
+
+        // If at least one task failed to start, re-throw exception and abort the start of the connector.
+        if (error != null) {
+            throw error;
+        }
+    }
+
+    /**
+     * Schedules polling of provided tasks and wait until all polling tasks eventually finish.
+     *
+     * @param tasks {@link List<EngineSourceTask>} of source tasks which should poll for the records.
+     */
+    private void runTasksPolling(final List<EngineSourceTask> tasks)
+            throws ExecutionException {
+        LOGGER.debug("Starting tasks polling.");
+        final ExecutorCompletionService<Void> taskCompletionService = new ExecutorCompletionService(taskService);
+        for (EngineSourceTask task : tasks) {
+            final RecordProcessor processor = selectRecordProcessor();
+            processor.initialize(recordService, transformations, new SourceRecordCommitter(task));
+            taskCompletionService.submit(new PollRecords(task, processor, state));
+        }
+
+        for (int i = 0; i < tasks.size(); i++) {
+            try {
+                taskCompletionService.take().get();
```

Comment:
I am not happy with the blocking call here. Can we change the loop so `poll()` can be used with pause between iterations?

Reply:
The thread is not in blocking state, but it should be in waiting state as `take()` [calls](https://github.com/openjdk/jdk/blob/be7cc1c2b083ac7cbcec6b0fe77caff16f14bb60/src/java.base/share/classes/java/util/concurrent/LinkedBlockingQueue.java#L435) `Conditon.await()`. So I'm not sure if there would be any advantage if we sleep here in a loop. Actually we would be just spending cycles here most of the time by awaking and sleeping threads with no effect.